segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,523 @@
+                                    .thenApply(this::deserializeEvents);
+        }
+
+        @SneakyThrows(Exception.class)
+        private EventsReadAndTruncationLength deserializeEvents(BufferView inputData) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long truncationLength = 0;
+            Exception deserializationException = new NoDataAvailableException();
+            try {
+                @Cleanup
+                BoundedInputStream input = new BoundedInputStream(inputData.getReader(), inputData.getLength());
+                int dataLength = inputData.getLength();
+                while (input.getRemaining() > 0 && events.size() < this.config.getMaxItemsAtOnce()) {
+                    ProcessorEventData event = this.serializer.deserialize(input);
+                    events.add(event);
+                    // Update the truncation length, which includes user data and internal VersionedSerializer metadata.
+                    truncationLength = dataLength - input.getRemaining();
+                }
+            } catch (BufferView.Reader.OutOfBoundsException ex) {
+                // Events are of arbitrary size, so it is quite possible we stopped reading in the middle of an event.
+                // Only if we have already read some data, we do not throw.
+            } catch (Exception ex) {
+                // Some unexpected serialization error occurred here; rethrow if this prevents reading further events.
+                log.error("{}: Problem while deserializing events.", this.traceObjectId);
+                deserializationException = ex;
+            }
+
+            // If no events have been read, throw the right exception. Otherwise, process whatever events have been read.
+            if (events.isEmpty()) {
+                throw deserializationException;
+            }
+            return new EventsReadAndTruncationLength(events, truncationLength);
+        }
+
+        private CompletableFuture<EventsReadAndTruncationLength> applyProcessorHandler(EventsReadAndTruncationLength readResult) {
+            return this.handler.apply(readResult.getProcessorEventsData()).thenApply(v -> readResult);
+        }
+
+        private CompletableFuture<Void> truncateInternalSegment(EventsReadAndTruncationLength readResult, Timer iterationTime) {
+            long truncationOffset = this.segment.getInfo().getStartOffset() + readResult.getTruncationLength();
+            return this.segment.truncate(truncationOffset, containerOperationTimeout)
+                               .thenAccept(v -> this.lastIterationLatency.set(iterationTime.getElapsedMillis()));
+        }
+
+        /**
+         * Utility class to group all events reads and provide some convenience methods to operate on them (e.g., bytes
+         * read accounting for headers).
+         */
+        @Data
+        class EventsReadAndTruncationLength {
```

Comment:
Please make all nested classes `static` unless you absolutely need something from the parent class' instance. Non-static inner classes always have a pointer to the parent class and are thus less memory efficient (and can lead to leaks if you hand their instances to the outside).

Reply:
Done.